vj4/handler/domain.py

```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
     self.json_or_redirect(self.url)
 
 
+@app.route('/domain/join_applications', 'domain_manage_join_applications')
+class DomainJoinApplicationsHandler(base.Handler):
+  @base.require_perm(builtin.PERM_EDIT_DESCRIPTION)
+  async def get(self):
+    roles = sorted(list(self.domain['roles'].keys()))
+    roles_with_text = [(role, role) for role in roles]
+    join_settings = domain.get_join_settings(self.domain)
+    expirations = [(k, v) for k, v in vj4.constant.domain.JOIN_EXPIRATION_RANGE.items()]
+    if join_settings is None:
+      expirations = expirations[1:]
+    self.render('domain_manage_join_applications.html', roles_with_text=roles_with_text,
+                join_settings=join_settings, expirations=expirations)
+
+  @base.require_perm(builtin.PERM_EDIT_DESCRIPTION)
+  @base.post_argument
+  @base.require_csrf_token
+  @base.sanitize
+  async def post(self, *, method: int, role_assignment: str=None, expire: int=None,
+                 invitation_code: str=''):
+    current_join_settings = domain.get_join_settings(self.domain)
+    if method not in constant.domain.JOIN_METHOD_RANGE:
+      raise error.ValidationError('method')
+    if method == constant.domain.JOIN_METHOD_NONE:
+      join_settings = None
+    else:
+      if role_assignment not in self.domain['roles']:
+        raise error.ValidationError('role_assignment')
+      if expire not in constant.domain.JOIN_EXPIRATION_RANGE:
+        raise error.ValidationError('expire')
+      if current_join_settings is None and expire == constant.domain.JOIN_EXPIRATION_KEEP_CURRENT:
+        raise error.ValidationError('expire')
+      if method == constant.domain.JOIN_METHOD_CODE:
+        validator.check_domain_invitation_code(invitation_code)
+      join_settings={'method': method, 'role': role_assignment}
+      if method == constant.domain.JOIN_METHOD_CODE:
+        join_settings['code'] = invitation_code
+      if expire == constant.domain.JOIN_EXPIRATION_KEEP_CURRENT:
+        join_settings['expire'] = current_join_settings['expire']
+      elif expire == constant.domain.JOIN_EXPIRATION_UNLIMITED:
+        join_settings['expire'] = None
+      else:
+        join_settings['expire'] = datetime.datetime.utcnow() + datetime.timedelta(hours=expire)
+    await domain.edit(self.domain_id, join=join_settings)
+    self.json_or_redirect(self.url)
+
+
+@app.route('/join', 'domain_join')
+class DomainJoinHandler(base.Handler):
+  async def ensure_user_not_member(self):
+    dudoc = await domain.get_user(self.domain_id, self.user['_id'])
+    if dudoc != None and 'role' in dudoc:
+      raise error.UserAlreadyDomainMemberError(self.domain_id, self.user['_id'])
+
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, code: str=''):
+    join_settings = domain.get_join_settings(self.domain)
+    if join_settings is None:
+      raise error.DomainJoinForbiddenError(self.domain_id)
+    await self.ensure_user_not_member()
```

Comment:
Check this before join_settings

Reply:
Seems that they are similar. both make sense.